File: backend/app/core/project_mode.py
from typing import Any

from app.core.dialogue import (
    SCRIPT_MODE_CUSTOM,
    SCRIPT_MODE_DIALOGUE_SCRIPT,
    SCRIPT_MODE_DUO_PODCAST,
    SCRIPT_MODE_SINGLE,
    resolve_script_mode,
)

VIDEO_MODE_ORAL_SCRIPT_DRIVEN = "oral_script_driven"
VIDEO_MODE_AUDIO_VISUAL_DRIVEN = "audio_visual_driven"
SUPPORTED_VIDEO_MODES = {
    VIDEO_MODE_ORAL_SCRIPT_DRIVEN,
    VIDEO_MODE_AUDIO_VISUAL_DRIVEN,
}

VIDEO_TYPE_CUSTOM = "custom"
VIDEO_TYPE_SINGLE_NARRATION = "single_narration"
VIDEO_TYPE_DUO_PODCAST = "duo_podcast"
VIDEO_TYPE_DIALOGUE_SCRIPT = "dialogue_script"

ORAL_SCRIPT_DRIVEN_VIDEO_TYPES = {
    VIDEO_TYPE_CUSTOM,
    VIDEO_TYPE_SINGLE_NARRATION,
    VIDEO_TYPE_DUO_PODCAST,
    VIDEO_TYPE_DIALOGUE_SCRIPT,
}


def resolve_video_mode(value: Any, default: str = VIDEO_MODE_ORAL_SCRIPT_DRIVEN) -> str:
    normalized = str(value or "").strip().lower()
    if normalized in SUPPORTED_VIDEO_MODES:
        return normalized
    return default


def resolve_video_type(
    value: Any,
    *,
    video_mode: str = VIDEO_MODE_ORAL_SCRIPT_DRIVEN,
    default: str = VIDEO_TYPE_CUSTOM,
) -> str:
    normalized = str(value or "").strip().lower()
    resolved_video_mode = resolve_video_mode(video_mode)
    if resolved_video_mode == VIDEO_MODE_AUDIO_VISUAL_DRIVEN:
        return VIDEO_TYPE_CUSTOM
    if normalized in ORAL_SCRIPT_DRIVEN_VIDEO_TYPES:
        return normalized
    return default


def resolve_script_mode_from_video_type(video_type: Any) -> str:
    resolved_video_type = resolve_video_type(video_type)
    if resolved_video_type == VIDEO_TYPE_DUO_PODCAST:
        return SCRIPT_MODE_DUO_PODCAST
    if resolved_video_type == VIDEO_TYPE_DIALOGUE_SCRIPT:
        return SCRIPT_MODE_DIALOGUE_SCRIPT
    if resolved_video_type == VIDEO_TYPE_CUSTOM:
        return SCRIPT_MODE_CUSTOM
    return SCRIPT_MODE_SINGLE


def resolve_video_type_from_script_mode(script_mode: Any) -> str:
    resolved_script_mode = resolve_script_mode(script_mode)
    if resolved_script_mode == SCRIPT_MODE_CUSTOM:
        return VIDEO_TYPE_CUSTOM
    if resolved_script_mode == SCRIPT_MODE_DUO_PODCAST:
        return VIDEO_TYPE_DUO_PODCAST
    if resolved_script_mode == SCRIPT_MODE_DIALOGUE_SCRIPT:
        return VIDEO_TYPE_DIALOGUE_SCRIPT
    return VIDEO_TYPE_SINGLE_NARRATION

File: frontend/src/lib/project-mode.ts
export type VideoMode = 'oral_script_driven' | 'audio_visual_driven'
export type VideoType = 'custom' | 'single_narration' | 'duo_podcast' | 'dialogue_script'
export type ScriptMode = 'custom' | 'single' | 'duo_podcast' | 'dialogue_script'

export const VIDEO_MODE_LABELS: Record<VideoMode, string> = {
  oral_script_driven: '口播文案驱动',
  audio_visual_driven: '声画驱动',
}

export const VIDEO_TYPE_LABELS: Record<VideoType, string> = {
  custom: '自定义',
  single_narration: '单人叙述',
  duo_podcast: '双人播客',
  dialogue_script: '台词剧本',
}

export function resolveVideoMode(value: unknown): VideoMode {
  const normalized = String(value || '').trim()
  if (normalized === 'audio_visual_driven') return 'audio_visual_driven'
  return 'oral_script_driven'
}

export function resolveVideoType(value: unknown): VideoType {
  const normalized = String(value || '').trim()
  if (normalized === 'custom') return 'custom'
  if (normalized === 'single_narration') return 'single_narration'
  if (normalized === 'duo_podcast') return 'duo_podcast'
  if (normalized === 'dialogue_script') return 'dialogue_script'
  return 'custom'
}

export function resolveScriptModeFromVideoType(value: unknown): ScriptMode {
  const type = resolveVideoType(value)
  if (type === 'duo_podcast') return 'duo_podcast'
  if (type === 'dialogue_script') return 'dialogue_script'
  if (type === 'custom') return 'custom'
  return 'single'
}

export function buildProjectModeLabel(modeValue: unknown, typeValue: unknown): string {
  const mode = resolveVideoMode(modeValue)
  const type = resolveVideoType(typeValue)
  return `${VIDEO_MODE_LABELS[mode]}-${VIDEO_TYPE_LABELS[type]}`
}
